core/services/blockhashstore/feeder.go

```diff
@@ -0,0 +1,164 @@
+) *Feeder {
+	return &Feeder{
+		logger:         logger,
+		coordinator:    coordinator,
+		bhs:            bhs,
+		waitBlocks:     waitBlocks,
+		lookbackBlocks: lookbackBlocks,
+		latestBlock:    latestBlock,
+		stored:         make(map[uint64]struct{}),
+		lastRunBlock:   0,
+	}
+}
+
+// Feeder checks recent VRF coordinator events and stores any blockhashes for blocks within
+// waitBlocks and lookbackBlocks that have unfulfilled requests.
+type Feeder struct {
+	logger         logger.Logger
+	coordinator    Coordinator
+	bhs            BHS
+	waitBlocks     int
+	lookbackBlocks int
+	latestBlock    func(ctx context.Context) (uint64, error)
+
+	stored       map[uint64]struct{}
+	lastRunBlock uint64
+}
+
+// Run the feeder.
+func (f *Feeder) Run(ctx context.Context) {
+	latestBlock, err := f.latestBlock(ctx)
+	if err != nil {
+		f.logger.Errorw("Failed to fetch current block number", "error", err)
+		return
+	}
+
+	var (
+		fromBlock        = int(latestBlock) - f.lookbackBlocks
+		toBlock          = int(latestBlock) - f.waitBlocks
+		blockToRequests  = make(map[uint64]map[string]struct{})
+		requestIDToBlock = make(map[string]uint64)
+	)
+	if fromBlock < 0 {
+		fromBlock = 0
+	}
+	if toBlock < 0 {
+		toBlock = 0
+	}
+	reqs, err := f.coordinator.Requests(ctx, uint64(fromBlock), uint64(toBlock))
+	if err != nil {
+		f.logger.Errorw("Failed to fetch VRF requests",
+			"error", err,
+			"latestBlock", latestBlock,
+			"fromBlock", fromBlock,
+			"toBlock", toBlock)
+		return
+	}
+	for _, req := range reqs {
+		if _, ok := blockToRequests[req.Block]; !ok {
+			blockToRequests[req.Block] = make(map[string]struct{})
+		}
+		blockToRequests[req.Block][req.ID] = struct{}{}
+		requestIDToBlock[req.ID] = req.Block
+	}
+
+	fuls, err := f.coordinator.Fulfillments(ctx, uint64(fromBlock))
+	if err != nil {
+		f.logger.Errorw("Failed to fetch VRF fulfillments",
+			"error", err,
+			"latestBlock", latestBlock,
+			"fromBlock", fromBlock,
+			"toBlock", toBlock)
+		return
+	}
+	for _, ful := range fuls {
+		requestBlock, ok := requestIDToBlock[ful.ID]
+		if !ok {
+			continue
+		}
+		delete(blockToRequests[requestBlock], ful.ID)
+	}
+
+	for block, unfulfilledReqs := range blockToRequests {
+		if len(unfulfilledReqs) > 0 {
+			if _, ok := f.stored[block]; ok {
+				// Already stored
+				continue
+			}
+			stored, err := f.bhs.IsStored(ctx, block)
+			if err != nil {
+				f.logger.Errorw("Failed to check if block is already stored, attempting to store anyway",
+					"error", err,
+					"block", block)
+			} else if stored {
+				f.logger.Infow("Blockhash already stored",
+					"block", block, "latestBlock", latestBlock)
+				f.stored[block] = struct{}{}
+				continue
+			}
+
+			// Block needs to be stored
+			err = f.bhs.Store(ctx, block)
+			if err != nil {
+				f.logger.Errorw("Failed to store block", "error", err, "block", block)
+				continue
+			}
+			f.logger.Infow("Stored blockhash",
+				"block", block, "latestBlock", latestBlock)
```

Comment:
Maybe a good idea to put `unfulfilledReqs` as a field to this log, would give context on all the requests that should be safe to execute a fulfillment for. In LogDNA we can just put the request ID in the search bar and it searches all the fields, so this log would show up if it has it.

Reply:
Done, I limited it to only 50 request IDs so we don't get massive logs